### PR TITLE
feat(): next

### PR DESCRIPTION
**Pull request by dvoegelin**
## **Description**

feat(QueryBuilder): New QueryBuilder component for dynamically generating queries #1328 

#### **Verify that...**

- [x] Any related demos were added and `npm start` and `npm run build` still works
- [x] New demos work in `Safari`, `Chrome` and `Firefox`
- [x] `npm run lint` passes
- [x] `npm test` passes and code coverage is increased
- [x] `npm run build` still works

#### **Bullhorn Internal Developers**
- [x] Run `Novo Automation`

##### **Screenshots**
176255941-8c8242f9-115f-433f-8e92-e483283c3225

### CODE REVIEW

File: projects/novo-elements/src/elements/field/formats/date-range-format.ts [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+    @Optional() @Inject(COMPOSITION_BUFFER_MODE) _compositionMode: boolean,
+    private labels: NovoLabelService,
+  ) {
+    super(_element, _renderer, _factory, _compositionMode);
+    const dateRangeFormat = this.labels.dateFormat.toUpperCase();
+
+    this.unmask = false;
+    this.imask = {
+      mask: 'm{/}`d{/}`Y - m{/}`d{/}`Y',
+      // pattern: 'm{/}`d{/}`Y - m{/}`d{/}`Y',
```
Comment (ndickerson) [inline projects/novo-elements/src/elements/field/formats/date-range-format.ts:45]: commented out code

File: projects/novo-elements/src/elements/field/formats/time-format.ts [hunk 2]
```diff
@@ -53,12 +53,12 @@ export class NovoTimeFormatDirective extends IMaskDirective<any> implements Novo
     private _element: ElementRef,
     _renderer: Renderer2,
     _factory: IMaskFactory,
-    @Optional() @Inject(PLATFORM_ID) _platformId: string,
+    // @Optional() @Inject(PLATFORM_ID) _platformId: string,
```
Comment (ndickerson) [inline projects/novo-elements/src/elements/field/formats/time-format.ts:56]: commented out code

File: projects/novo-elements/src/elements/query-builder/condition-builder/condition-builder.component.html [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+    </div>
+
+    <div class="condition-input">
+      <ng-container conditionInputOutlet></ng-container>
+    </div>
+  </novo-grid>
+  <ng-content></ng-content>
+</form>
+
+<!-- EMPTY STATE TEMPLATE -->
```
Comment (ndickerson) [inline projects/novo-elements/src/elements/query-builder/condition-builder/condition-builder.component.html:40]: commented out code - do we need this or not?

File: projects/novo-elements/src/elements/query-builder/condition-builder/condition-builder.component.scss [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+  .and-or-filter-button {
+    box-sizing: border-box;
+    background: #fff;
+    border: 1px solid #ddd;
+    color: #5691f5;
+    display: inline-block;
+    position: relative;
+    cursor: pointer;
+    margin: 0.4rem auto;
+    // vertical-align: middle;
```
Comment (ndickerson) [inline projects/novo-elements/src/elements/query-builder/condition-builder/condition-builder.component.scss:56]: commented out code

File: projects/novo-elements/src/elements/query-builder/condition-builder/condition-builder.component.ts [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+    this.parentForm.valueChanges.subscribe((value) => {
+      Promise.resolve().then(() => this.onFieldSelect());
+    });
+  }
+
+  ngAfterContentInit() {
+    const { fields = [] } = this.qbs.config || {};
+    fields.length && this.changeFieldOptions(fields[0]);
+    this.searches = this.searchTerm.valueChanges.pipe(debounceTime(300), distinctUntilChanged()).subscribe((term) => {
+      // this.parentForm.get('field').setValue(null);
```
Comment (ndickerson) [inline projects/novo-elements/src/elements/query-builder/condition-builder/condition-builder.component.ts:88]: commented out code

File: projects/novo-elements/src/elements/query-builder/condition-builder/condition-builder.component.ts [hunk 6]
```diff
@@ -0,0 +1,203 @@
...
+    }
+  }
+
+  ngOnDestroy() {
+    this.searches.unsubscribe();
+    // Clear all outlets and Maps
+    [this._operatorOutlet.viewContainer, this._inputOutlet.viewContainer].forEach((def) => {
+      def.clear();
+    });
+    // this._contentFieldTypeDefs = [];
```
Comment (ndickerson) [inline projects/novo-elements/src/elements/query-builder/condition-builder/condition-builder.component.ts:110]: commented out code

File: projects/novo-elements/src/elements/query-builder/condition-builder/condition-builder.component.ts [hunk 7]
```diff
@@ -0,0 +1,203 @@
...
+    this._onDestroy.next();
+    this._onDestroy.complete();
+  }
+
+  /**
+   * Updates the Conditions "Field" Options to Change base on new Scope
+   * @param fieldConfig
+   */
+  changeFieldOptions(fieldConfig: FieldConfig<BaseFieldDef>) {
+    // this.fields = entity.filter(term);
```
Comment (ndickerson) [inline projects/novo-elements/src/elements/query-builder/condition-builder/condition-builder.component.ts:121]: commented out code

File: projects/novo-elements/src/elements/query-builder/condition-group/condition-group.component.html [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+          </novo-button>
+        </novo-flex>
+      </ng-container>
+    </ng-container>
+    <button theme="dialogue" data-automation-id="add-advanced-search-condition" icon="add-thin" side="left" size="sm" uppercase padding="sm" (click)="addCondition()">
+      {{ labels.addCondition }}</button>
+  </novo-stack>
+
+
+  <!-- <button class="and-or-button" theme="secondary" size="sm" (click)="addRootCondition()">{{ addCriteriaLabel }}</button> -->
```
Comment (ndickerson) [inline projects/novo-elements/src/elements/query-builder/condition-group/condition-group.component.html:30]: commented out code

File: projects/novo-elements/src/elements/query-builder/criteria-builder/criteria-builder.component.html [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+<novo-number-condition-def name="FLOAT"></novo-number-condition-def>
+<novo-number-condition-def name="INTEGER"></novo-number-condition-def>
+<novo-number-condition-def name="BIGDECIMAL"></novo-number-condition-def>
+<novo-number-condition-def name="DOUBLE"></novo-number-condition-def>
+<novo-address-condition-def name="ADDRESS"></novo-address-condition-def>
+<novo-boolean-condition-def name="BOOLEAN"></novo-boolean-condition-def>
+<novo-picker-condition-def name="SELECT"></novo-picker-condition-def>
+<novo-string-condition-def name="DEFAULT"></novo-string-condition-def>
+
+<!--
```
Comment (ndickerson) [inline projects/novo-elements/src/elements/query-builder/criteria-builder/criteria-builder.component.html:23]: commented out code

File: projects/novo-examples/src/components/query-builder/just-criteria/just-criteria-example.ts [hunk 10]
```diff
@@ -0,0 +1,157 @@
...
+      {
+        scope: 'Candidate',
+        field: 'customDate1',
+        operator: 'between',
+        value: {
+          startDate: '1922-01-01',
+          endDate: '1922-01-17',
+        }
+      },
+      // where=category IN (1,2,3)
```
Comment (ndickerson) [inline projects/novo-examples/src/components/query-builder/just-criteria/just-criteria-example.ts:140]: commented out code

**Review by ndickerson — COMMENTED** [review]
Noted where there is commented out code. Not sure if this is known, and planned to fix later or not.

File: projects/novo-elements/src/elements/chips/ChipList.ts [hunk 11]
```diff
@@ -516,7 +519,7 @@ export class NovoChipList
     if (this._chipInput && this._chipInput.focused) {
       // do nothing
     } else if (this._chipInput) {
-      this._focusInput(options);
+      Promise.resolve().then(() => this._focusInput(options));
```
Comment (ccarbonellbh) [inline projects/novo-elements/src/elements/chips/ChipList.ts:522]: What does this change give us here?

File: projects/novo-elements/src/elements/field/formats/date-format.ts [hunk 12]
```diff
@@ -34,11 +28,11 @@ export class NovoDateFormatDirective extends IMaskDirective<any> {
     private _element: ElementRef,
     _renderer: Renderer2,
     _factory: IMaskFactory,
-    @Optional() @Inject(PLATFORM_ID) _platformId: string,
+    // @Optional() @Inject(PLATFORM_ID) _platformId: string,
```
Comment (ccarbonellbh) [inline projects/novo-elements/src/elements/field/formats/date-format.ts:31]: Can we delete?

File: projects/novo-elements/src/elements/field/formats/time-format.ts (hunk 2 above)
Comment (ccarbonellbh) [inline projects/novo-elements/src/elements/field/formats/time-format.ts:56]: Can we delete?

File: projects/novo-elements/src/elements/query-builder/condition-definitions/string-condition.definition.ts [hunk 13]
```diff
@@ -0,0 +1,83 @@
+import { COMMA, ENTER } from '@angular/cdk/keycodes';
+import { ChangeDetectionStrategy, Component, ViewEncapsulation } from '@angular/core';
+import { AbstractControl, FormControl } from '@angular/forms';
+import { AbstractConditionFieldDef } from './abstract-condition.definition';
+
+/**
+ * Contruction filters against String fields can be complex. Each "chip" added to the
+ * condition can be used to indendantly used to query a database.  Not all systems support
```
Comment (ccarbonellbh) [inline projects/novo-elements/src/elements/query-builder/condition-definitions/string-condition.definition.ts:8]: independently?

File: projects/novo-examples/src/updates/june-2022.md [hunk 14]
```diff
@@ -0,0 +1,35 @@
...
+page: v7-query-builder
+title: June 2022
+order: 1
+tag: new
+---
+
+📢  June 2022 (version 7.2.x)
+===========================
+
+**New Feature: Query Builder**: We have added a new component to help with building complex and modern interfaces.  At Bullhorn we have to enable our users to create advanced and dynamic searches against there data.  We have built many iterations of a query builder in the past but we think this one is our best and most re-usable.
```
Comment (ccarbonellbh) [inline projects/novo-examples/src/updates/june-2022.md:12]:
```suggestion
**New Feature: Query Builder**: We have added a new component to help with building complex and modern interfaces.  At Bullhorn we have to enable our users to create advanced and dynamic searches against their data.  We have built many iterations of a query builder in the past but we think this one is our best and most re-usable.
```

**Review by ccarbonellbh — APPROVED** [review]

**Pull request merged**